### PR TITLE
Refactor view

### PR DESCRIPTION
**Pull request by davidbrochart**

### CODE REVIEW

File: include/zarray/zarray.hpp [hunk 1]
```diff
@@ -94,6 +94,8 @@ namespace xt
         const nlohmann::json& get_metadata() const;
         void set_metadata(const nlohmann::json& metadata);
 
+        zarray_impl* make_view(xstrided_slice_vector& slices);
```
Comment (JohanMabille) [inline include/zarray/zarray.hpp:97]: THis method should return a `zarray` object wrapping the `zarray_impl*` returned by the `make_view` method of `zarray_impl`

File: include/zarray/zarray_impl.hpp [hunk 2]
```diff
@@ -430,6 +435,14 @@ namespace xt
         return new self_type(*this);
     }
 
+    template <class CTE>
+    inline zarray_impl* zexpression_wrapper<CTE>::make_view(xstrided_slice_vector& slices)
+    {
+        auto e = strided_view(m_expression, slices);
+        zarray_impl* p_impl = e.derived_cast().allocate_result();
```
Comment (JohanMabille) [inline include/zarray/zarray_impl.hpp:442]: `e.derived_cast()` is an xexpression that do not provide the `allocate_result` method. What you want to do here is build a `zexpression_wrapper` on the newly created strided_view:

```cpp
return build_zarray(std::move(e));
```

File: include/zarray/zarray_impl.hpp [hunk 3]
```diff
@@ -536,6 +549,14 @@ namespace xt
         return new self_type(*this);
     }
 
+    template <class CTE>
+    inline zarray_impl* zscalar_wrapper<CTE>::make_view(xstrided_slice_vector& slices)
+    {
+        auto e = strided_view(m_array, slices);
+        zarray_impl* p_impl = e.derived_cast().allocate_result();
```
Comment (JohanMabille) [inline include/zarray/zarray_impl.hpp:556]: same remark as previous

File: include/zarray/zarray_impl.hpp [hunk 4]
```diff
@@ -656,6 +677,14 @@ namespace xt
         return new self_type(*this);
     }
 
+    template <class CTE>
+    inline zarray_impl* zarray_wrapper<CTE>::make_view(xstrided_slice_vector& slices)
+    {
+        auto e = strided_view(m_array, slices);
+        zarray_impl* p_impl = e.derived_cast().allocate_result();
```
Comment (JohanMabille) [inline include/zarray/zarray_impl.hpp:684]: Same remark here

File: include/zarray/zarray_impl.hpp [hunk 5]
```diff
@@ -756,6 +785,14 @@ namespace xt
         return new self_type(*this);
     }
 
+    template <class CTE>
+    inline zarray_impl* zchunked_wrapper<CTE>::make_view(xstrided_slice_vector& slices)
+    {
+        auto e = strided_view(m_chunked_array, slices);
+        zarray_impl* p_impl = e.derived_cast().allocate_result();
```
Comment (JohanMabille) [inline include/zarray/zarray_impl.hpp:792]: Same remark here

**Review by JohanMabille — CHANGES_REQUESTED** [review]

**Comment by davidbrochart** [comment]
I couldn't use `detail::is_strided_view` because [it already exists](https://github.com/xtensor-stack/xtensor/blob/6846c156825356a62dc16193245936988793fe6e/include/xtensor/xview.hpp#L158-L164) (but it doesn't seem to be what we need here). So I named it `detail::is_a_strided_view` :smile:

File: include/zarray/zarray_impl.hpp [hunk 6]
```diff
@@ -17,9 +17,85 @@
...
+    template <class CTE>
+    class zchunked_wrapper;
+
+    template <class CTE>
+    class zexpression_wrapper;
+
+    namespace detail
+    {
+        template <class E>
+        struct is_a_strided_view : std::false_type
```
Comment (JohanMabille) [inline include/zarray/zarray_impl.hpp:37]: Nitpicking: could be `is_xstrided_view` to be consistent with `is_xarray` just below

File: include/zarray/zarray_impl.hpp [hunk 7]
```diff
@@ -17,9 +17,85 @@
...
+        {
+        };
+
+        template <class E, class S>
```
Comment (JohanMabille) [inline include/zarray/zarray_impl.hpp:41]: I think you could add the [additional template arguments](https://github.com/xtensor-stack/xtensor/blob/master/include/xtensor/xstrided_view.hpp#L125) of `xstrided_view` to be sure we can match all the cases.

File: include/zarray/zarray_impl.hpp [hunk 8]
```diff
@@ -431,24 +499,24 @@ namespace xt
...
+        return strided_view_impl(slices, detail::is_a_strided_view<CTE>());
     }
 
     template <class CTE>
-    inline std::size_t zexpression_wrapper<CTE>::get_offset() const
+    inline zarray_impl* zexpression_wrapper<CTE>::strided_view_impl(xstrided_slice_vector& slices, std::true_type)
     {
         compute_cache();
-        return m_cache.get_offset();
+        auto e = xt::strided_view(m_cache, slices);
```
Comment (JohanMabille) [inline include/zarray/zarray_impl.hpp:511]: What about:

```cpp
auto e = xt::strided_view(get_array(), slices);
```

This way you're not exposed to the caching mechanism detail.

**Review by JohanMabille — COMMENTED** [review]
I think the file `zview.hpp`  and the [zview_dispatcher](https://github.com/xtensor-stack/zarray/blob/master/include/zarray/zdispatcher.hpp#L51) can be removed.

**Review by JohanMabille — APPROVED** [review]

**Comment by JohanMabille** [comment]
Great! I'll merge when the CI is green.

**Pull request merged**